### PR TITLE
feat: retry mechanism

### PR DESCRIPTION
**Pull request by yevheniyJ**

### CODE REVIEW

File: src/core/internal/retry/index.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+export class RetryService {
+    constructor(private config: RetryConfig) {}
+
+    /**
+     *
+     * @param func function to execute (should return {@link promise})
+     */
+    async executeAsyncFunc(func: Function): Promise<any> {
+        for (let i = 0; i < this.config.retries; i++) {
```
Comment (alexandrtovmach) [inline src/core/internal/retry/index.ts]:
```suggestion
        for (let i = 0; i <= this.config.retries; i++) {
```

File: src/core/internal/retry/index.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            }
+        }
+    }
+
+    /**
+     *
+     * @param func function to execute
+     */
+    async executeSyncFunc(func: Function): Promise<any> {
+        for (let i = 0; i < this.config.retries; i++) {
```
Comment (alexandrtovmach) [inline src/core/internal/retry/index.ts]:
```suggestion
        for (let i = 0; i <= this.config.retries; i++) {
```

File: src/core/internal/retry/index.ts (hunk 1 above)
Comment (yevheniyJ) [reply src/core/internal/retry/index.ts]: Well it’s really depends on how we count retries. My solution was based on that even first (standard) execution is a try. But I changed my mind and your proposal seems to be more logical :) first execution is a regular execution. And overall (maxim) number of possible invocations will be retries + 1

Comment (yevheniyJ) [reply src/core/internal/retry/index.ts]: fixed

**Pull request merged**